include/sisl/fds/compact_bitset.hpp

```diff
@@ -0,0 +1,147 @@
+    void reset_bit(bit_count_t start) { set_reset_bit(start, false); }
+
+    bool is_bit_set(bit_count_t bit) const {
+        bitword_type const* word_ptr = get_word_const(bit);
+        if (!word_ptr) { return false; }
+        uint8_t const offset = get_word_offset(bit);
+        return word_ptr->is_bit_set_reset(offset, true);
+    }
+
+    bit_count_t get_next_set_bit(bit_count_t start_bit) const { return get_next_set_or_reset_bit(start_bit, true); }
+    bit_count_t get_next_reset_bit(bit_count_t start_bit) const { return get_next_set_or_reset_bit(start_bit, false); }
+
+    void set_reset_bit(bit_count_t bit, bool value) {
+        bitword_type* word_ptr = get_word(bit);
+        if (!word_ptr) { return; }
+        uint8_t const offset = get_word_offset(bit);
+        word_ptr->set_reset_bits(offset, 1, value);
+    }
+
+    bit_count_t get_next_set_or_reset_bit(bit_count_t start_bit, bool search_for_set_bit) const {
+        bit_count_t ret{inval_bit};
+
+        // check first word which may be partial
+        uint8_t const offset = get_word_offset(start_bit);
+        bitword_type const* word_ptr = get_word_const(start_bit);
+        if (!word_ptr) { return ret; }
+
+        uint8_t nbit{0};
+        bool found = search_for_set_bit ? word_ptr->get_next_set_bit(offset, &nbit)
+                                        : word_ptr->get_next_reset_bit(offset, &nbit);
+        if (found) { ret = start_bit + nbit - offset; }
+
+        if (ret == inval_bit) {
+            // test rest of whole words
+            bit_count_t current_bit = start_bit + (bitword_type::bits() - offset);
+            bit_count_t bits_remaining = (current_bit > size()) ? 0 : size() - current_bit;
+            while (bits_remaining > 0) {
+                ++word_ptr;
+                found =
+                    search_for_set_bit ? word_ptr->get_next_set_bit(0, &nbit) : word_ptr->get_next_reset_bit(0, &nbit);
+                if (found) {
+                    ret = current_bit + nbit;
+                    break;
+                }
+                current_bit += bitword_type::bits();
+                bits_remaining -= std::min< bit_count_t >(bits_remaining, bitword_type::bits());
+            }
+        }
+
+        if (ret >= size()) { ret = inval_bit; }
+        return ret;
+    }
+
+    std::string to_string() const {
+        std::string str;
+        auto const num_words = size() / word_size_bytes();
+        for (uint32_t i{0}; i < num_words; ++i) {
+            fmt::format_to(std::back_inserter(str), "{}", s_->words[i].to_string());
+        }
+        return str;
+    }
+
+private:
+    bitword_type* get_word(bit_count_t bit) {
+        return (sisl_unlikely(bit >= nbits_)) ? nullptr : &s_->words[bit / word_size_bits()];
+    }
+
+    bitword_type const* get_word_const(bit_count_t bit) const {
+        return (sisl_unlikely(bit >= nbits_)) ? nullptr : &s_->words[bit / word_size_bits()];
+    }
+
+    uint8_t get_word_offset(bit_count_t bit) const {
```

Comment:
How about: `get_offset_in_word`? it took me a while to understand what this api is for until I read upto the the return value is consumed by wordptr. :-)

Reply:
I didn't implement this now, but took from bitset.hpp. I didn't want to spend time on reimplement them and maintain correctness etc and wanted to piggy back on bitset.hpp, which is battle tested.